crates/revm/src/optimism/mod.rs

```diff
@@ -0,0 +1,176 @@
+    // + 32 bytes for the block hash
+    // + 32 bytes for the block sequence number
+    // + 32 bytes for the batcher hash
+    // + 32 bytes for the fee overhead
+    // + 32 bytes for the fee scalar
+    if data.len() != 256 {
+        return Err(reth_executor::BlockExecutionError::OptimismBlockExecution(
+            reth_executor::OptimismBlockExecutionError::L1BlockInfoError {
+                message: "unexpected l1 block info tx calldata length found".to_string(),
+            },
+        ))
+    }
+
+    let l1_base_fee = U256::try_from_be_slice(&data[64..96]).ok_or(
+        reth_executor::BlockExecutionError::OptimismBlockExecution(
+            reth_executor::OptimismBlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 base fee".to_string(),
+            },
+        ),
+    )?;
+    let l1_fee_overhead = U256::try_from_be_slice(&data[192..224]).ok_or(
+        reth_executor::BlockExecutionError::OptimismBlockExecution(
+            reth_executor::OptimismBlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 fee overhead".to_string(),
+            },
+        ),
+    )?;
+    let l1_fee_scalar = U256::try_from_be_slice(&data[224..256]).ok_or(
+        reth_executor::BlockExecutionError::OptimismBlockExecution(
+            reth_executor::OptimismBlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 fee scalar".to_string(),
+            },
+        ),
+    )?;
+
+    Ok(L1BlockInfo { l1_base_fee, l1_fee_overhead, l1_fee_scalar })
+}
+
+/// An extension trait for [L1BlockInfo] that allows us to calculate the L1 cost of a transaction
+/// based off of the [ChainSpec]'s activated hardfork.
+pub trait RethL1BlockInfo {
+    /// Forwards an L1 transaction calculation to revm and returns the gas cost.
+    ///
+    /// ### Takes
+    /// - `chain_spec`: The [ChainSpec] for the node.
+    /// - `timestamp`: The timestamp of the current block.
+    /// - `input`: The calldata of the transaction.
+    /// - `is_deposit`: Whether or not the transaction is a deposit.
+    fn l1_tx_data_fee(
+        &self,
+        chain_spec: Arc<ChainSpec>,
+        timestamp: u64,
+        input: &Bytes,
+        is_deposit: bool,
+    ) -> Result<U256, BlockExecutionError>;
+
+    /// Computes the data gas cost for an L2 transaction.
+    ///
+    /// ### Takes
+    /// - `chain_spec`: The [ChainSpec] for the node.
+    /// - `timestamp`: The timestamp of the current block.
+    /// - `input`: The calldata of the transaction.
+    fn l1_data_gas(
+        &self,
+        chain_spec: Arc<ChainSpec>,
+        timestamp: u64,
+        input: &Bytes,
+    ) -> Result<U256, BlockExecutionError>;
+}
+
+impl RethL1BlockInfo for L1BlockInfo {
+    fn l1_tx_data_fee(
+        &self,
+        chain_spec: Arc<ChainSpec>,
+        timestamp: u64,
+        input: &Bytes,
+        is_deposit: bool,
+    ) -> Result<U256, BlockExecutionError> {
```

Comment:
do these need an `Arc` or would `&ChainSpec` be sufficient?

Reply:
Updated to use referenced chain specs (`&ChainSpec`) since downstream usage referenced object-owned data.

Addressed in https://github.com/anton-rs/op-reth/pull/155